logbook-core/src/main/java/org/zalando/logbook/XmlCompactingBodyFilter.java

```diff
@@ -0,0 +1,74 @@
+package org.zalando.logbook;
+
+import lombok.extern.slf4j.Slf4j;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+import javax.annotation.Nullable;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathFactory;
+import java.io.ByteArrayInputStream;
+import java.io.StringWriter;
+import java.util.function.Predicate;
+
+import static javax.xml.transform.OutputKeys.INDENT;
+import static javax.xml.transform.OutputKeys.OMIT_XML_DECLARATION;
+import static javax.xml.xpath.XPathConstants.NODESET;
+import static org.zalando.fauxpas.FauxPas.throwingSupplier;
+
+@Slf4j
+class XmlCompactingBodyFilter implements BodyFilter {
+
+    private final Predicate<String> contentTypes = MediaTypeQuery.compile("*/xml", "*/*+xml");
+    private final Transformer transformer = transformerFactory();
+
+    @Override
+    public String filter(@Nullable final String contentType, final String body) {
+        return contentTypes.test(contentType) ? compact(body) : body;
+    }
+
+    private String compact(final String body) {
+        if (body.trim().isEmpty()) return body;
+        try {
+            final StringWriter output = new StringWriter();
+            final Document document = documentWithoutTextNodes(body);
+            transformer.transform(new DOMSource(document), new StreamResult(output));
+            return output.toString();
+        } catch (Exception e) {
+            log.trace("Unable to compact body, is it a XML?. Keep it as-is: `{}`", e.getMessage());
+            return body;
+        }
+    }
+
+    private Document documentWithoutTextNodes(final String body) {
```

Comment:
This seems rather inefficient, a custom sax or stax parser should have much less overhead.

There are also some security implication when parsing the request payload, see <https://en.wikipedia.org/wiki/XML_external_entity_attack> and <https://www.owasp.org/index.php/XML_External_Entity_(XXE)_Prevention_Cheat_Sheet#JAXP_DocumentBuilderFactory.2C_SAXParserFactory_and_DOM4J> for how to prevent this in java.

Reply:
I would agree with a fact, that sax/stax parsing will be more effective than dom in general.
Unfortunately, potential benefit from introducing custom parser with supporting rules will be miserable, due to fact of parsing already allocated in memory string .

Reply:
I had applied recommended changes, according to XXE